### PR TITLE
Episode 1: Add expression evaluator

### PR DESCRIPTION
**Pull request by terrajobst**
This adds the basics we need, specifically a lexer, a parser, and diagnostics that will lay the foundation for our compiler.

### CODE REVIEW

File: docs/episode-01.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Episode 1
+
+[Video](https://www.youtube.com/watch?v=wgHIkdUQbp0) |
```
Comment (jnm2) [inline docs/episode-01.md]: This link contains the exact start time: https://youtu.be/wgHIkdUQbp0?t=9m44s

Comment (terrajobst) [reply docs/episode-01.md]: I thought about doing that but instead I've trimmed the video on YouTube. Unfortunately it seems to take them forever to process that :-(

Comment (jnm2) [reply docs/episode-01.md]: Great, that's better! Thanks for letting me know. I'll remove my "Starts at" comment on YouTube so as not to confuse folks. 😃

**Comment by mattwar** [comment]
Are you going to merge this?

File: mc/CodeAnalysis/Lexer.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            _text = text;
+        }
+
+        public IEnumerable<string> Diagnostics => _diagnostics;
+
+        private char Current
+        {
+            get
+            {
+                if (_position >= _text.Length)
```
Comment (LindaLawton) [inline mc/CodeAnalysis/Lexer.cs]: why dont you test for a less than 0 _position here?  are you not worried about an out of range exception?

Comment (terrajobst) [reply mc/CodeAnalysis/Lexer.cs]: Because I know that never decrement — that would be extremely weird.

**Review by Tobii24 — COMMENTED** [review]
.

File: mc/CodeAnalysis/Lexer.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+                while (char.IsWhiteSpace(Current))
+                    Next();
+
+                var length = _position - start;
+                var text = _text.Substring(start, length);
+                return new SyntaxToken(SyntaxKind.WhitespaceToken, start, text, null);
+            }
+
+            if (Current == '+')
```
Comment (rokiboxofficial) [inline mc/CodeAnalysis/Lexer.cs:70]: Wouldn't it be better to use a dictionary with SyntaxToken by not numeric symbol?

**Pull request merged**